source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,591 @@
+                    // Normalize direction vector
+                    if (direction.norm() > 0.0)
+                      direction *= 1./direction.norm();
+                    // TODO this is only correct for box geometries
+                    else
+                      direction[dim-1] = 1.;
+
+                    // Compute the total displacement in the gravity direction,
+                    // i.e. the initial topography + any additional mesh displacement.
+                    const double displacement = direction * (displacement_values[point] + initial_topography_values[point]);
+
+                    // To project onto the tangent space of the surface,
+                    // we define the projection P:= I- n x n,
+                    // with I the unit tensor and n the unit normal to the surface.
+                    // The surface gradient then is P times the usual gradient of the shape functions.
+                    const Tensor<2, dim, double> projection = unit_symmetric_tensor<dim>() -
+                                                              outer_product(fs_fe_face_values.normal_vector(point), fs_fe_face_values.normal_vector(point));
+
+                    // Loop over the shape functions
+                    for (unsigned int i=0; i<dofs_per_cell; ++i)
+                      {
+                        // Make sure we only assemble for the y-component
+                        // TODO this is only correct for box geometries
+                        if (mesh_deformation_dof_handler.get_fe().system_to_component_index(i).first != dim-1)
+                          continue;
+
+                        // Assemble the RHS
+                        // RHS = M*H_old
+                        cell_vector(i) += displacement *
+                                          fs_fe_face_values.shape_value (i, point) *
+                                          fs_fe_face_values.JxW(point);
+
+
+                        for (unsigned int j=0; j<dofs_per_cell; ++j)
+                          {
+                            // Make sure we only assemble for the y-component
+                            // TODO this is only correct for box geometries
+                            if (mesh_deformation_dof_handler.get_fe().system_to_component_index(j).first != dim-1)
+                              continue;
+                            // Assemble the matrix, for backward first order time discretization:
+                            // Matrix := (M+dt*K) = (M+dt*B^T*kappa*B)
+                            cell_matrix(i,j) +=
+                              (
+                                this->get_timestep() * diffusivity *
+                                (projection * fs_fe_face_values.shape_grad(i, point)) *
+                                (projection * fs_fe_face_values.shape_grad(j, point)) +
+                                fs_fe_face_values.shape_value (i, point) * fs_fe_face_values.shape_value (j, point)
```

Comment:
Since you have `M+dt*...`, move this line to the front of the formula to match the comment